Extend docs about project initialization (#1785)
software-mansion/protostar: website/docs/tutorials/08-cairo-1-support/02-project-initialization.md

Reviewer: This file at the moment is very chaotic and contains as well the info on `protostar.toml`
This page should be broken down into at least two:
- Modules/Packages
- Project initialization
- (and perhaps protostar.toml)?

After running the command, the following structure will be generated:

```
my_project/
├── hello_starknet/
│   ├── src/
│   │   ├── business_logic/
│   │   │   └── utils.cairo
│   │   ├── contracts/
│   │   │   └── hello_starknet.cairo
│   │   ├── business_logic.cairo

> │   │   ├── contracts.cairo

Reviewer 2: Can users have multiple contracts in contracts? If not we shouldn't use plural

Reviewer 2: It it is there to be future proof, we should add an explicit  warning

Author: You cannot have multiple contracts anywhere within the crate. I'll add a warning.

│   │   └── lib.cairo
│   └── cairo_project.toml
├── tests/
│   ├── test_hello_starknet.cairo
│   └── test_utils.cairo
└── protostar.toml
```

:::warning
This template will be changed in future versions, but the old one will still be usable with newer protostar versions
:::

### `hello_starknet`

> This directory contains our only crate in this project - `hello_starknet`.

Reviewer: ```suggestion
This directory contains our only module in this project - `hello_starknet`.
```

Reviewer: This is confusing, we don't know a difference between those (is there any difference?)

### `cairo_project.toml` and `lib.cairo`

All Cairo1 crates must define these files.

- `cairo_project.toml` - this file defines the name of the crate. It is good practice for this name to match the
  top-level directory name - `hello_starknet` in our case.

> - `lib.cairo` - this file exposes all of our modules to the compiler. Initially, it only contains two modules:

Reviewer 2: ```suggestion
- `lib.cairo` - this fil all of our modules we wish to include in the library. Initially, it only contains two modules:
```

Author: I think it's more confusing the way you wrote it.

Reviewer: doesn't make much sense :P

```cairo title="lib.cairo"
mod business_logic;
mod contracts;
```

You can learn about [modules](#cairo-1-modules) and [how to add them](#adding-a-new-module) in further sections.

### `contracts`

> This directory contains the code of our contract - `HelloStarknet`. For easy testing, we recommend this directory

Reviewer 2: ```suggestion
This directory contains the code of our contract - `HelloStarknet`. As a good practice, we recommend this directory
```

contains

> only the contract definition that utilizes functions imported from other modules.

Reviewer 2: ```suggestion
only the contract definition, bussines logic should be kept in other modules.
```

### `business_logic`

This directory contains standalone cairo1 methods that can be imported and used in the contract definition. We recommend
writing business logic in this directory for ease of writing unit tests.

### `contracts.cairo` and `business_logic.cairo`

> `contracts.cairo` and `business_logic.cairo`: All modules (directories) must be "imported" to the top-level directory.

Reviewer 2: ```suggestion
`contracts.cairo` and `business_logic.cairo`: All modules must be listed in `lib.cairo`
```

> Note that these files only contain `mod ...` definitions that import files from directories.

Reviewer 2: ```suggestion
```
This sentence is a bit confusing

> Creation of these files is necessary so that they can be imported in the `lib.cairo file`.

Reviewer 2: Isn't this too obvious information?

Author: I'll simplify this a bit, but I think it's still good info to have for the completness sake.

```cairo title="contracts.cairo
mod hello_starknet;
```

### `tests`

> All tests should be defined in this directory.

Reviewer 2: Can we link testing section?

### `protostar.toml`

> This file contains the configuration for the Protostar project.

Reviewer 2: Can we link cairo1 toml section?

│   └── test_utils.cairo
└── protostar.toml
```

#### Using added module

You now use your function in the HelloStarknet contract use `my_project::mod1::functions::returns_three()`.

## The protostar.toml

Apart from the usual things you can find in `protostar.toml`, there is a `linked-libarires` entry which is used to find
cairo1 modules in tests and building.
This makes it possible for you to include other modules from your dependencies if they are correct cairo1 modules (with
their own module definition and `cairo_project.toml`).

```
[project]
protostar-version = "0.9.2"
lib-path = "lib"
linked-libraries = ["hello_starknet"]

[contracts]
hello_starknet = ["hello_starknet"]
```

:::info
Even though `hello_starknet.cairo` file is defined in the nested directory, we use `"src"` as path to the contract. This
is necessary for the imports from other modules (like `business_logic`) to work.
:::

## Using multiple contracts in project

Due to limitations of cairo1 compiler, having multiple contracts defined in the project will cause `protostar build` and
other commands to fail.

**That is, having projects structured like this is not valid and will not work correctly.**

> ### ❌ Incorrect multi-contract project structure

Reviewer 2: We shouldn't include incorrect schema, I don't see a point in that

Author: I think for some users having an example of what cannot be done may be useful. Especially since I think many will initially attempt to create a structure like this.

Reviewer: i disagree, no point in showing wrong structure

After running the command, the following structure will be generated:

```
my_project/
├── hello_starknet/
│   ├── src/
│   │   ├── business_logic/
│   │   │   └── utils.cairo
│   │   ├── contracts/
│   │   │   └── hello_starknet.cairo
│   │   ├── business_logic.cairo
│   │   ├── contracts.cairo
│   │   └── lib.cairo
│   └── cairo_project.toml
├── tests/
│   ├── test_hello_starknet.cairo
│   └── test_utils.cairo
└── protostar.toml
```

:::warning
This template will be changed in future versions, but the old one will still be usable with newer protostar versions
:::

### `hello_starknet`

This directory contains our only package in this project - `hello_starknet`.

### `cairo_project.toml` and `lib.cairo`

All Cairo1 packages must define these files.

> You can learn about [packages](./01-understanding-cairo-packages.md) and how

Reviewer 2: ❤️

to [add new module to a package](./01-understanding-cairo-packages.md#adding-a-new-module) in
further sections.

### `contracts`

This directory contains the code of our contract - `HelloStarknet`. As a good practice, we recommend this directory
contains only the contract definition, business logic should be kept in other modules.

:::danger
Currently protostar only supports having one contract per package. You cannot add more contracts to this directory. To
use multiple contracts in your project see [this section](#using-multiple-contracts-in-project).
:::

### `business_logic`

This directory contains standalone cairo1 methods that can be imported and used in the contract definition. We recommend

> writing business logic in this directory for ease of writing unit tests.

Reviewer 2: ```suggestion
writing business logic in this directory to simplify writing unit tests.
```

### `contracts.cairo` and `business_logic.cairo`

These files are necessary so that they can be imported in the `lib.cairo` file.

### `tests`

All [tests](./04-testing.md) should be defined in this directory.

### `protostar.toml`

This file contains the [configuration for the Protostar project](./03-protostar-toml.md).

:::info
Even though `hello_starknet.cairo` file is defined in the nested directory, we use a package
directory `"hello_starknet"` as path to the contract. This is necessary for the imports from modules within package
containing the contract (like `business_logic`) to work.
:::

## Using multiple contracts in project

> Due to limitations of cairo1 compiler, having multiple contracts defined in the project will cause `protostar build` and

Reviewer 2: ```suggestion
Due to limitations of cairo1 compiler, having multiple contracts defined in the package will cause `protostar build` and
```

> Due to limitations of cairo1 compiler, having multiple contracts defined in the project will cause `protostar build` and
> other commands to fail.

Reviewer 3: ```suggestion
Due to limitations of the cairo1 compiler, having multiple contracts defined in the project will cause a `protostar build` and
other commands to fail.
```

├── package2/
│   ├── src/
│   │   ├── contracts/
│   │   │   └── other_contract.cairo
│   │  ...
│   │   ├── contracts.cairo
│   │   └── lib.cairo
│   └── cairo_project.toml
...
└── protostar.toml
```

Make sure `[crate_roots]` are correctly defined.

```toml title="package1/cairo_project.toml"
[crate_roots]
package1 = "src"
```

```toml title="package2/cairo_project.toml"
[crate_roots]
package2 = "src"
```

Define each contract in the `[contracts]` section of the protostar.toml and each package
in the `linked-librares`

```toml title="protostar.toml"
# ...
linked-libraries = ["package1", "package2"]

[contracts]
hello_starknet = ["package1"]
other_contract = ["package2"]
```

### Testing multi-contract projects

For example, to test function `returns_two` defined in the `package1/business_logic/utils.cairo` write

```cairo title="Example test"
#[test]
fn test_returns_two() {
    assert(package1::business_logic::utils::returns_two() == 2, 'Should return 2');
}
```

Or using the `use path:to::mod` syntax

```cairo title="Example test
use package1::business_logic::utils::returns_two;

#[test]
fn test_returns_two() {
    assert(returns_two() == 2, 'Should return 2');
}
```

> Make sure that the path::to:the::module is correct for your package structure.

Reviewer 2: Can we add info how to do integration tests as well?

Author: What do you mean exactly? Perhaps let's just link to the testing page of the docs here so we don't duplicate knowledge.

Reviewer 2: That's fine

After running the command, the following structure will be generated:

```
my_project/
├── hello_starknet/
│   ├── src/
│   │   ├── business_logic/
│   │   │   └── utils.cairo
│   │   ├── contracts/
│   │   │   └── hello_starknet.cairo
│   │   ├── business_logic.cairo
│   │   ├── contracts.cairo
│   │   └── lib.cairo
│   └── cairo_project.toml
├── tests/
│   ├── test_hello_starknet.cairo
│   └── test_utils.cairo
└── protostar.toml
```

:::warning
This template will be changed in future versions, but the old one will still be usable with newer protostar versions
:::

### `hello_starknet`

This directory contains our only package in this project - `hello_starknet`.

### `cairo_project.toml` and `lib.cairo`

All Cairo1 packages must define these files.

- `cairo_project.toml` - this file defines the name of the package. It is good practice for this name to match the
  top-level directory name - `hello_starknet` in our case.

> - `lib.cairo` - this file exposes all of our package to the compiler. Initially, it only contains two modules:

Reviewer: you're kind of re-explaining them here. Instead i would suggest linking to the explanation and merging this explanation with the one from `understanding-cairo-packages`

and `lib.cairo` files defined.

```
my_project/
├── package1/
│   ├── src/
│   │   ├── contracts/
│   │   │   └── hello_starknet.cairo
│   │  ...
│   │   ├── contracts.cairo
│   │   └── lib.cairo
│   └── cairo_project.toml
├── package2/
│   ├── src/
│   │   ├── contracts/
│   │   │   └── other_contract.cairo
│   │  ...
│   │   ├── contracts.cairo
│   │   └── lib.cairo
│   └── cairo_project.toml
...
└── protostar.toml
```

Make sure `[crate_roots]` are correctly defined.

```toml title="package1/cairo_project.toml"
[crate_roots]
package1 = "src"
```

```toml title="package2/cairo_project.toml"
[crate_roots]
package2 = "src"
```

Define each contract in the `[contracts]` section of the protostar.toml and each package
in the `linked-librares`

```toml title="protostar.toml"
# ...
linked-libraries = ["package1", "package2"]

[contracts]
hello_starknet = ["package1"]
other_contract = ["package2"]
```

### Testing multi-contract projects

You can write your tests in the standard manner. Make sure you use correct paths.
For example, to test function `returns_two` defined in the `package1/business_logic/utils.cairo` write

```cairo title="Example test"
#[test]
fn test_returns_two() {

>     assert(package1::business_logic::utils::returns_two() == 2, 'Should return 2');

Reviewer: we should `use` to show that testing can be less cumbersome
